### PR TITLE
add Geocentric and Local Cartesian projectors

### PR DESCRIPTION
**Pull request by mantkiew**
The UTM projector creates shifts in projected maps relative to actual world positions.

Added Geocentric (ECEF) and LocalCartesian projectors from geographic lib as available projectors.

These projectors have the advantage over the UTM projector in that they properly use lat/lon/alt to compute XYZ using WGS84 ellipsoid, whereas the UTM projector only uses lat/lon to convert into X/Y/Z on the ellipsoid and just adds elevation, which is incorrect.

image

In the image above, the point P' is the shifted point created by the UTM projector, compared to the correct point P.

### CODE REVIEW

**Comment by m-naumann** [comment]
@mantkiew thanks! Briefly browsing through: would you please add tests?

**Comment by mantkiew** [comment]
@m-naumann I added the tests and mentioned the new projectors in the README and CHANGELOG as well. 

BTW. We are using these extensions/fixes from my PRs in this project:

https://uwaterloo.ca/waterloo-intelligent-systems-engineering-lab/news/avin-lanelet2-mapping-project-completed

Additionally, the WISE ADS now also depends on these projectors

https://uwaterloo.ca/waterloo-intelligent-systems-engineering-lab/projects/wise-automated-driving-system

We previously used Lanelet1 but we're now migrating to use Lanelet2.

**Comment by m-naumann** [comment]
> @m-naumann I added the tests and mentioned the new projectors in the README and CHANGELOG as well.
> 
> BTW. We are using these extensions/fixes from my PRs in this project:
> 
> https://uwaterloo.ca/waterloo-intelligent-systems-engineering-lab/news/avin-lanelet2-mapping-project-completed
> 
> Additionally, the WISE ADS now also depends on these projectors
> 
> https://uwaterloo.ca/waterloo-intelligent-systems-engineering-lab/projects/wise-automated-driving-system
> 
> We previously used Lanelet1 but we're now migrating to use Lanelet2.

Thanks a lot! And nice to see the usage. You'll certainly not regret the update to lanelet2 ;) And at least I personally can back your call for correct elevation consideration.

Could you also add tests for the python bindings (even though we didn't provide them initially, we should try to increase as much covered code as possible).

And please bear with us for a bit, we need to get a public CI pipeline up and running again, such that static code analysis is available on github and there is no longer a need to report issues from the internal pipeline.

**Comment by mantkiew** [comment]
@m-naumann  I have added the tests. Note that the values coming back from the projectors have to be rounded for comparison. The maximum number of decimals that work is `8`. The tests pass, of course.

I believe that is ready to be merged.

**Comment by mantkiew** [comment]
@m-naumann on the second thought: maybe we should round the values to the eight decimal ourselves in the API? That would take away the burden from the user.

**Comment by m-naumann** [comment]
> And please bear with us for a bit, we need to get a public CI pipeline up and running again, such that static code analysis is available on github and there is no longer a need to report issues from the internal pipeline.

This is now done, @mantkiew would you please merge master into your branch to have the pipeline running for this PR?

**Comment by mantkiew** [comment]
Wonderful! Glad to finish this up. On it....

**Comment by mantkiew** [comment]
@m-naumann All GitHub actions have successfully completed in my repository. You have to allow them here.

**Comment by mantkiew** [comment]
@m-naumann the test job on noetic fails with some nonsense `Error: sha is not defined`. Could you please restart it? All jobs passed on my repository (see https://github.com/wiselabuw/Lanelet2/actions/runs/3281889493)

**Comment by immel-f** [comment]
> @m-naumann the test job on noetic fails with some nonsense `Error: sha is not defined`. Could you please restart it? All jobs passed on my repository (see https://github.com/wiselabuw/Lanelet2/actions/runs/3281889493)

This is a bug in the new CI, once [#259](https://github.com/fzi-forschungszentrum-informatik/Lanelet2/pull/259) is merged it should also pass here.

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class LocalCartesianProjector : public Projector {
+ public:
+  explicit LocalCartesianProjector(Origin origin);
+
+  BasicPoint3d forward(const GPSPoint& gps) const override;
+
+  GPSPoint reverse(const BasicPoint3d& enu) const override;
+
+ private:
+  GeographicLib::LocalCartesian localCartesian;
```
Comment (poggenhans) [inline lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:19]: should be `localCartesian_`, to conform to our convention for private members

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include <lanelet2_io/Exceptions.h>
```
Comment (poggenhans) [inline lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:2]: unused include

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include <lanelet2_io/Projection.h>
+
+#include <GeographicLib/Geocentric.hpp>
```
Comment (poggenhans) [inline lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:5]: should also be moved to the cpp file

**Review by poggenhans — CHANGES_REQUESTED** [review]
I'm adding a few comments. Sorry for taking so long. Once these points are resolved, we can finally merge this PR.

File: lanelet2_projection/CHANGELOG.rst [hunk 4]
```diff
@@ -2,6 +2,12 @@
 Changelog for package lanelet2_projection
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+1.1.2 (2022-03-18)
```
[the remark this answers is not in the document]
Comment (mantkiew) [reply lanelet2_projection/CHANGELOG.rst:5]: Done, replaced with the version from `master`.

File: lanelet2_projection/include/lanelet2_projection/Geocentric.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <lanelet2_io/Exceptions.h>
+#include <lanelet2_io/Projection.h>
+
+#include <GeographicLib/Geocentric.hpp>
```
[the remark this answers is not in the document]
Comment (mantkiew) [reply lanelet2_projection/include/lanelet2_projection/Geocentric.h:5]: done

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h (hunk 1 above)
Comment (mantkiew) [reply lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:19]: renamed

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h (hunk 2 above)
Comment (mantkiew) [reply lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:2]: removed

File: lanelet2_projection/include/lanelet2_projection/LocalCartesian.h (hunk 3 above)
Comment (mantkiew) [reply lanelet2_projection/include/lanelet2_projection/LocalCartesian.h:5]: moved

**Comment by mantkiew** [comment]
@poggenhans ok, this is now ready to go.

File: lanelet2_projection/CHANGELOG.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (poggenhans) [reply lanelet2_projection/CHANGELOG.rst]: Thanks :-). Can you change the version  in the package.xml too?

[the remark this answers is not in the document]
Comment (mantkiew) [reply lanelet2_projection/CHANGELOG.rst]: done

**Review by poggenhans — APPROVED** [review]

**Comment by mantkiew** [comment]
@poggenhans hi, anything preventing this to be merged?

**Pull request merged**